### PR TITLE
Rework of Doctrine ItemDataProvider

### PR DESCRIPTION
**Pull request by theofidry**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | #457
| License       | MIT
| Doc PR        | none

The objective of this PR is to:

- Rework the code to make it a bit more explicit and declarative
- Tackle the concerns risen by #457

### CODE REVIEW

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php [hunk 1]
```diff
@@ -88,13 +147,35 @@ public function getItem(string $resourceClass, $id, string $operationName = null
...
+    private function retrieveQueryBuilder(ObjectManager $resourceManager, string $resourceClass)
+    {
+        $repository = $resourceManager->getRepository($resourceClass);
+
+        if ($repository instanceof  EntityRepository) {
+            return $repository->createQueryBuilder('o');
+        }
+
+        if ($resourceManager instanceof EntityManagerInterface) {
+            // TODO: check if $resourceClass is good here
```
Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: TODO to be checked: @dunglas you're probably more familiar than me on that one

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php [hunk 2]
```diff
@@ -88,13 +147,35 @@ public function getItem(string $resourceClass, $id, string $operationName = null
...
+            return $resourceManager->createQueryBuilder()
+                ->select('o')
+                ->from($resourceClass, 'o');
         }
 
-        $repository = $manager->getRepository($resourceClass);
-        $queryBuilder = $repository->createQueryBuilder('o');
+        // TODO: handle this case
```
Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: #457: second issue risen

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php [hunk 3]
```diff
@@ -32,22 +36,27 @@ class ItemDataProvider implements ItemDataProviderInterface
...
      */
-    public function __construct(ManagerRegistry $managerRegistry, PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, array $itemExtensions = [], ItemDataProviderInterface $decorated = null)
-    {
+    public function __construct(
+        ManagerRegistry $managerRegistry,
+        PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory,
+        PropertyMetadataFactoryInterface $propertyMetadataFactory,
+        array $itemExtensions = [],
+        ItemDataProviderInterface $decoratedProvider = null
+    ) {
```
Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: code style change: can't remember what was our policy here but I'm not found of having to scroll like crazy to see all the arguments... hence the change

Comment (dunglas) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: We follow Symfony CS: all on the same line.

Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: erf, fine let's have a 250 columns long line then :disappointed:

Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: Word wrap ;)

Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: too heavy with some files when working on the front-end^^

**Comment by theofidry** [comment]
In blocked state:

- waiting for input on https://github.com/api-platform/core/pull/458#discussion_r55929098
- waiting on https://github.com/phpspec/prophecy/pull/263 for tests

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php [hunk 4]
```diff
@@ -88,13 +147,35 @@ public function getItem(string $resourceClass, $id, string $operationName = null
...
+     * @param ObjectManager $resourceManager
+     * @param string        $resourceClass
+     *
+     * @return \Doctrine\ORM\QueryBuilder
+     */
+    private function retrieveQueryBuilder(ObjectManager $resourceManager, string $resourceClass)
+    {
+        $repository = $resourceManager->getRepository($resourceClass);
+
+        if ($repository instanceof  EntityRepository) {
```
Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: @dunglas should I check against `EntityRepository` or try if it has the `createQueryBuilder()` method?

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php (hunk 1 above)
Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: could also check against the method, not sure

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php (hunk 2 above)
Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: To myself: throw a `RunTime` exception

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php (hunk 4 above)
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: Prefer `class_exists` it will allow custom implementations to work too.

**Comment by theofidry** [comment]
ping @dunglas updated the querybuilder part, tell me if it's good for you and I'll add unit tests afterwards

**Comment by soyuka** [comment]
@theofidry I made a use case that broke in v1 due to composite identifiers, feel free to use the [behat test case I wrote](https://github.com/api-platform/core/pull/387/commits/d49e08ec61e9d9c35a83e6e604084ec2cf05f7f1) :).

**Comment by theofidry** [comment]
@soyuka neat, will look into it as soon as I can :)

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php [hunk 5]
```diff
@@ -154,24 +152,42 @@ private function getIdentifiers(string $resourceClass, $id)
...
     private function retrieveQueryBuilder(ObjectManager $resourceManager, string $resourceClass)
     {
         $repository = $resourceManager->getRepository($resourceClass);
 
-        if ($repository instanceof  EntityRepository) {
+        if ($repository instanceof EntityRepository) {
             return $repository->createQueryBuilder('o');
         }
 
         if ($resourceManager instanceof EntityManagerInterface) {
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: Can't you add an `or` condition to avoid code duplication below? (same goes for the repository `method_exists`).

Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: not really because the order matter here:

- try with an `EntityRepository`: optimal case
- try with a `EntityManagerInterface`: other optimal case, but requires to do things differently
- fallback to check if the repository still have this `createQueryBuilder()` method, and assumes it's the right one (technically there is no guarantee here)
- last fallback with the manager before throwing an error

Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: I find this hack pretty ugly but maybe I missed something, it's been some time since I used Doctrine, would be happy to find a better solution

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: Hmm I don't understand why `EntityManagerInterface` would take priority upon `EntityRepository` if it has a `createQueryBuilder` method.

```
        if ($repository instanceof EntityRepository || method_exists($repository, 'createQueryBuilder')) {
            return $repository->createQueryBuilder('o');
        }

        if ($resourceManager instanceof EntityManagerInterface || method_exists($resourceManager, 'createQueryBuilder')) {
            return $resourceManager->createQueryBuilder()
                ->select('o')
                ->from($resourceClass, 'o');
        }

        throw new RuntimeException();
```  

I'd add that I find it weird to test that the `$resourceManager` has the `createQueryBuilder`, especially if it's not implementing the `EntityManagerInterface`. IMO this will almost always be false, what's the edge case that I'm missing?

IMO there is no better way of returning a query builder with a select clause on the correct entity, and it's a fine way of doing it.

Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: >Hmm I don't understand why EntityManagerInterface would take priority upon EntityRepository if it has a createQueryBuilder method.

My thinking was trying to get the most reliable solution first, assuming that the `EntityReopsitory::createQueryBuilder()` is right is a baseless assumption with no means to check if it's true. At least with the `EntityManagerInterface` it's guaranteed to work.

>I'd add that I find it weird to test that the $resourceManager has the createQueryBuilder, especially if it's not implementing the EntityManagerInterface. IMO this will almost always be false, what's the edge case that I'm missing?

`$resourceManager` could simply be a `ObjectManager` right?

IMO there is no better way of returning a query builder with a select clause on the correct entity, and it's a fine way of doing it.

I don't have any strong feeling on this, but if you and the others find it better I would be happy to switch to that instead

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: > $resourceManager could simply be a ObjectManager right?

Indeed, and the `ObjectManager` doesn't provide the `createQueryBuilder` method. `EntityManagerInterface` extends `ObjectManager`, and the `createQueryBuilder` method belongs to the `EntityManagerInterface`. 

As I'm no Doctrine expert, I've no further opinion on this. Hope others will :).

Comment (theofidry) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: Well, we got someone who wrote a book on it right @dunglas?

**Comment by theofidry** [comment]
ping @dunglas

**Comment by theofidry** [comment]
@api-platform/core-team there is one last point for which I'm not sure of the behaviour we should adopt, see [the discussion](https://github.com/api-platform/core/pull/458#discussion_r57160917) with soyuka. Could you check that so that I fix this and rebase this PR?

**Comment by theofidry** [comment]
ping @api-platform/core-team: there is just one part left, could you guys check it please?

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php [hunk 6]
```diff
@@ -15,59 +15,116 @@
...
-    private $decorated;
+    private $decoratedProvider;
 
     /**
      * @param ManagerRegistry                        $managerRegistry
      * @param PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory
      * @param PropertyMetadataFactoryInterface       $propertyMetadataFactory
      * @param QueryItemExtensionInterface[]          $itemExtensions
-     * @param ItemDataProviderInterface|null         $decorated
+     * @param ItemDataProviderInterface|null         $decoratedProvider
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: We're naming this `decorated` in every other class, why change it here?

File: src/Bridge/Doctrine/Orm/ItemDataProvider.php (hunk 5 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/ItemDataProvider.php]: I've read one more time and it does make sense in some way :).

**Comment by theofidry** [comment]
Updated

**Comment by theofidry** [comment]
@dunglas is that still relevant? i.e. is it still worth to rebase it?

**Comment by dunglas** [comment]
@theofidry Honestly I don't know. WDYT?

**Comment by theofidry** [comment]
Let me check it when I'm back in London (Tuesday), any opinion before that
would not be very reasonable ^^

**Comment by soyuka** [comment]
Maybe we can close this one?

**Comment by theofidry** [comment]
I don't really have time to take a look at it again so yeah let's do that. The problem may not be solved but I don't see any API change so that's something that could easily come as a patch if a problem were to occur.
